### PR TITLE
Module - 3: File System, Streams, HTTP

### PR DESCRIPTION
**Pull request by idutka**

### CODE REVIEW

**Review by lionvs — APPROVED** [review]
Overall looks great!

We can also think about some issues with the current `csv-file` approach/implementaiton.

1. When we query events of a specific location we still are loading all events in our memory. Potentially we can have millions of all events but only a few of them with the needed location.

2. If we have concurrent requests for POST/DELETE endpoints we can lose some data.
For example (R1- request 1, R2-request2, ):
R1 create event with id 5
R2 create event with id 6
R1 reads data from file
R2 reads data from file
R1 adds event with id 5 to the data
R2 adds event with id 6 to the data (note that event with id 5 is not there)
If we now query all events we could see that event with id 5 is missing.

**Comment by idutka** [comment]
@lionvs Please take a look at some improvements:

1. Reading events by a stream for a specific location
2. Lock file for POST/PUT/DELETE request
3. POST request appends new Event at the end of the file 

What was not implemented:
1. Validation
2. Response with error status code

File: module-3/utils.js [hunk 1]
```diff
@@ -0,0 +1,160 @@
+const fs = require('fs');
+const path = require('path');
+const stream = require('stream');
+const csvParser = require('csv-parse');
+const lockfile = require('proper-lockfile');
+
+const filePath = path.join('.', 'events.csv');
+
+const lockCsv = () => lockfile.lock(filePath, {
```
Comment (lionvs) [inline module-3/utils.js:9]: +1 For using locking!

File: module-3/utils.js [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    })
+    stream.write(getCsvItem(event) + '\n');
+    stream.end();
+  });
+}
+
+const getEvents = async () => {
+  return new Promise((resolve) => {
+    fs.readFile(filePath, (err, data) => {
+      if (err) throw err;
```
Comment (lionvs) [inline module-3/utils.js:54]: Instead of throw we may have to use `reject(err)` here since we are creating our own promise here.

**Review by lionvs — COMMENTED** [review]
Looks Great!

**Pull request merged**